### ISSUE TITLE
Crowdsale improvements

### ISSUE THREAD

**Issue by jakub-wojciechowski** [OpenZeppelin/zeppelin-solidity#310]
First of all, @frangio thx for the amazing [Crowdsale PR](https://github.com/OpenZeppelin/zeppelin-solidity/pull/283). I've really enjoyed reading the code. Great job :) !

While reading I've noted on a side a few comments and suggestions for improvements. I'd be grateful if you let me know what do you think about them:

1. Crowdsale contract is tightly coupled to MintableToken. Wouldn’t it be better to make the base Crowdsale as generic as possible? It will allow others to implement a model when the token pool is created before a sale and then transferred rather than minted. Currently, it’s impossible to override methods like:

`	function createTokenContract() internal returns (MintableToken) {`

I will suggest extracting:

`token.mint(beneficiary, tokens);`

to

```
 function distribute(beneficiary, tokens) internal {
  token.mint(beneficiary, tokens); 
}
```

And probably rename: 

`function createTokenContract() => setupTokenContract()`		
	
2. Precondition checks are kept within a method body mixing with business logic. It’s a better practice to [extract them to modifiers](https://medium.com/@gavofyork/condition-orientated-programming-969f6ba0161a):

```
require(beneficiary != 0x0);
require(validPurchase());
```

3. Balance update is very verbose:
```
uint256 weiAmount = msg.value;
uint256 updatedWeiRaised = weiRaised.add(weiAmount);

….

// update state
weiRaised = updatedWeiRaised;
```

Am I missing any benefit of such separation? Why not just:

`weiRaised = weiRaised.add(msg.value);`

4. Code that checks if a sale has ended is duplicated leading to the risk of inconsistency once updated or overridden:

`bool withinPeriod = current >= startBlock && current <= endBlock;`

```
function hasEnded() public constant returns (bool) {
	return block.number > endBlock;
}
```

5. **CappedCrowdsale**. Conditions that checks if a cap is reached are independently implemented in two places (similar to above):

```
bool withinCap = weiRaised.add(msg.value) <= cap;

bool capReached = weiRaised >= cap;
```

6. **CappedCrowdsale**. Missing sanity check for `cap > 0`

7. **RefundableCrowdsale**. Missing sanity check for `goal > 0`

8. **RefundVault**. Should we start in an `ACTIVE` state when `Crowdsale` is not activated yet. This may potentially lead to accepting money outside of `Crowdsale`. The vault lifecycle should correspond to the `Crowdsale` lifecycle.
 
9. **RefundVault** Once again we may replace inline checks with modifiers:

`require(state == State.Active); => inState(State.Active)`

10. **RefundVault**. Function:

`function refund(address investor) `

is open to everyone. I think we limit the privilege to depositors or owner. It will be surprising if a third party initiate refunds. Why not use a standard pullPayment pattern in here.
 
11. **RefundVault**. It will be nice to have a clear policy regulating what to do with untouched funds. Part of them may be stuck forever due to lost keys. We may introduce a  refund period after which the remaining money may be sent to the crowdsale wallet or split among all of the investors.

12. It will be nice to have well-defined states that describe Crowdsale lifecycle like: 

`enum State {CREATED, ACTIVE, PAUSED, CLOSED, FINALIZED}`

 It will add more clarity and reduce code duplication in checking functions preconditions. We may even introduce a [state machine pattern](http://solidity.readthedocs.io/en/develop/common-patterns.html#state-machine) so derived Crowdsale contracts may easily add code when certain transition happen (like `onTransition(ACTIVE,CLOSED)`.

13. The last comment is rather an invitation to discussion. What do you think about replacing inheritance with composition in the form of mixings (implemented as linked libraries)? My experience shows that this approach is usually more flexible and easier for maintenance. I’m just afraid that we may run into trouble when there are additional features requests and the inheritance hierarchy gets deeper. (Event currently there are 3 levels: Crowdsale -> FinalizableCrowdsale -> RefundableCrowdsale)

Thanks for reading that. I'm looking forward to hearing your opinions.

**Comment by frangio** [comment]
Thanks a lot @jakub-wojciechowski! The crowdsale contracts were a joint effort by the OpenZeppelin team 😄.

It's great to see you excited about collaborating. There's a lot of suggestions in this PR and I can't go over everything thoroughly at the moment but I will revisit it in the next couple of days and get back to you. Are you in the OpenZeppelin slack so I can reach you there?

In the meantime, though, there are a couple of suggestions which you can go ahead and open PRs for: 3, and the missing sanity checks.

**Comment by jakub-wojciechowski** [comment]
Cool, thanks for the instant reply. I will open the PRs. 
I've just joined the Slack as "kuba". I'm looking forward to discussing this stuff whenever you are ready.

**Comment by miohtama** [comment]
Some projects prefer "allocate + burn" token sales over "mint" token sales. In this case the whole token supply is created with the token contract. Then one gives `approve()` for the crowdsale contract and the crowdsale contract does `transferFrom()`.

https://github.com/TokenMarketNet/ico/blob/master/contracts/AllocatedCrowdsale.sol#L54

**Comment by jakub-wojciechowski** [comment]
I totally agree. This is exactly the scenario that I'm describing in the 1. comment. Happy to hear that you have similar views.

**Comment by frangio** [comment]
I forgot to comment on that, but yeah I completely agree that we should make Crowdsale more generic! We want to stay pragmatic though, and not abstract more than people need.

**Comment by miohtama** [comment]
For 2. Precondition oriented programming it's more about the coding style. In some point you hit the limits of Solidity language design. When you have a lot of precondition checks the function declarations grow longish. In this point you start to act against the original precondition oriented idea to make the code more readable. Furthermore in some cases it makes sense for preconditions happen in a certain order, after function has taken some preparation steps to calculate variables together.

But in the end, I feel it's a matter of taste and one really should not care as long as the project as a whole takes one approach and follows it from the start to the end.

**Comment by frangio** [comment]
- Moved 1 to #351.
- Moved 4 and 5 to #352.
- 3, 6 and 7 and have been fixed in #312.
- 2 and 9 we won't fix for the reasons @miohtama stated above. (The new `require` function makes preconditions in function bodies look much nicer.)
- 8 doesn't make much sense to me because the only address allowed to send money to the contract is the crowdsale, so there is no risk of someone sending money at a different time. Additionally, a specific crowdsale might have more stages in its lifecycle than the `RefundVault` might care, e.g. if there is a prefunding period the refund vault behaves the same.
- Regarding 11, I'd prefer to keep the simpler current behavior. I think we can expect people to want their money back, and do without the added complexity.
- 12 would be nice but it seems non trivial to do. I will think about it. Allowing inheriting contracts to intercept state transitions sounds very attractive.
- 13 is very broad but it's on my mind too. We tend to favor inheritance for reasons of gas costs and to make deployment simpler. We could open an issue about this or discuss it in slack.

I love that you went through the code and proposed enhancements and things to discuss about it! Next time make sure to open separate issues so that we can discuss things separately and take action independently. 😄
